src/modules/apps/add.ts

```diff
@@ -1,94 +1,74 @@
+import { ManifestEditor, ManifestValidator } from '../../lib/manifest'
 import log from '../../logger'
-import { getManifest, getManifestPath, namePattern, vendorPattern, wildVersionPattern } from '../../manifest'
-
-import { appLatestMajor, handleError, parseArgs, pickLatestVersion, wildVersionByMajor } from './utils'
+import { appLatestMajor, parseArgs, pickLatestVersion, wildVersionByMajor } from './utils'
 
-const unprefixName = compose<string, string[], string>(last, split(':'))
-const invalidAppMessage =
-  'Invalid app format, please use <vendor>.<name>, <vendor>.<name>@<version>, npm:<name> or npm:<name>@<version>'
+const unprefixName = (str: string) => {
+  return str.split(':').pop()
+}
 
-const infraLatestVersion = (app: string): Promise<string | never> =>
-  router
-    .getAvailableVersions(app)
-    .then<string[]>(path(['versions', region()]))
-    .then(pickLatestVersion)
-    .then(wildVersionByMajor)
-    .catch(handleError(app))
+const invalidAppMessage = 'Invalid app format, please use <vendor>.<name>, <vendor>.<name>@<version>'
 
-const npmLatestVersion = (app: string): Promise<string | never> => {
-  return latestVersion(app)
-    .then(concat('^'))
-    .catch(err => Promise.reject(new CommandError(err.message)))
-}
+const infraLatestVersion = async (app: string) => {
+  try {
+    const { versions } = await router.getAvailableVersions(app)
+    const latest = pickLatestVersion(versions[region()])
+    return wildVersionByMajor(latest)
+  } catch (err) {
+    if (err.response && err.response.status === 404) {
+      throw new CommandError(`App ${chalk.green(app)} not found`)
+    }
 
-const updateManifestDependencies = (app: string, version: string): Promise<void> => {
-  return getManifest()
-    .then((manifest: Manifest) => {
-      return {
-        ...manifest,
-        dependencies: {
-          ...manifest.dependencies,
-          [app]: version,
-        },
-      }
-    })
-    .then((newManifest: Manifest) => JSON.stringify(newManifest, null, 2) + '\n')
-    .then((manifestJson: string) => writeFile(getManifestPath(), manifestJson))
+    throw err
+  }
 }
 
-const addApp = (app: string): Promise<void> => {
+const addApp = async (app: string, manifest: ManifestEditor) => {
   const hasVersion = app.indexOf('@') > -1
   if (hasVersion) {
     const [appId, version] = app.split('@')
-    return updateManifestDependencies(appId, version)
+    return manifest.addDependency(appId, version)
   }
-  const isNpm = app.startsWith('npm:')
+
   const isInfra = app.startsWith('infra:')
   const appName = app.includes(':') ? unprefixName(app) : app
-  const versionRequest = isNpm
-    ? npmLatestVersion(appName)
-    : isInfra
-    ? infraLatestVersion(appName)
-    : appLatestMajor(appName)
-  return versionRequest.then((version: string) => updateManifestDependencies(app, version))
+  const version = await (isInfra ? infraLatestVersion(appName) : appLatestMajor(appName))
+
+  return manifest.addDependency(app, version)
```

Comment:
_nit_:  you can write something like:
```
const [appId, version] = app.split('@')
const sanitizedVersion = version ? version : getVersion(appName)
return manifest.addDependency(appId, sanitizedVersion)
```
and the `getVersion` function:
```
const getVersion = (appName: string) => {
    const isInfra = appName.startsWith('infra:')
    const name= appName.includes(':) ? unprefixName(appName) : appName
    return await (isInfra ? infraLatestVersion(name) : appLatestMajor(name))
}
```

In my mind it is a bit clearer since you call `manifest.addDependency` in only one place and also separate the `getVersion` part

Reply:
It looks cleaner and clearer :+1: